### ISSUE TITLE
[Shaders] Improved Framebuffer-Attachment organization

### ISSUE THREAD

**Issue by rre36** [sp614x/optifine#3194]
Exposing manual control about the framebuffer attachment binding to us shader developers could make the limit of only eight freely usable attachments a lot less restricting.
The idea would be that we have a file called ``framebuffer.properties`` or something similar where we can define a set of framebuffer attachements and then specifiy which one is being bound as what colortex sampler in a specific pass eg.:
```
# This would be defining a buffer with a buffer name and a format.
buffer.BufferName1 = rgba16f
buffer.BufferName2 = rgba16

# Being able to specify that this buffer should be non-clearing here would also be useful
# since temporal buffers don't have to "burn" an entire colortex for the whole pipeline anymore.
# Could look like this:
buffer.BufferName1.nonclearing = true

# This would be binding said buffer to a colortex in a specified pass.
deferred1.colortex2 = BufferName1

# Binding custom textures could also be done in a similar manner.
# This kind of binding would obviously have to be done for every pass
# but I think most of us shader developers would be willing to do this
# if we suffer from buffer shortage, especially since not every buffer is needed in every pass anyways.
buffer.BufferName2.texture="path to texture"
```

If the ``framebuffer.properties`` file does not exist optifine could fallback to the regular behavior.

**Comment by zombye** [comment]
So overall this is something I do think would be great to have, but there are a lot of things in the example that I think should be different.

* Instead of `nonclearing` it should just be `clear`.
There's no reason to have it be the inverse of how this currently works.
* Since in the example there can be actual names for each image, it'd be nice if those names could also be used as the names of the sampler uniforms.
So for example we could have something like this: `<program>.sampler.<index> = <imageName>`
And then the image could be accessed through `uniform sampler2D <imageName>;` in the shaders.
* Instead of using `buffer` for this I suggest using `image`.
In modern GLSL both of these terms are already used for specific things; but an `image` would be closer to what we're actually referring to.
I'm fine with either option, but I'll be using `image` for the rest of this comment.
* Instead of `<program>.Drawbuffer<index> = <imageName>`, I think it should be `<program>.drawbuffer.<index> = <imageName>`.
It's pretty similar, but I think it makes more sense.
* What about flipping?
Since it wasn't mentioned I have to assume it'd just be the same as before but apply to each image.
But with this system, flipping is no longer necessary since equivalent behavior would be easy (albeit a little tedious) to specify.
* I think defining the images themselves should probably be a bit different from the example.
I'm not really sure what a good way to do it would be, but to me it feels like how it worked in the example is a bad way to do it.
* Lastly, having the file be named `framebuffer.properties` feels wrong to me, buuut I don't really have a better suggestion either.

**Comment by rre36** [comment]
Yeah, I just wantend to outline how it could work. The stuff you mentioned definitly makes sense when talking about how it should be in the actual implementation.

**Comment by rre36** [comment]
Updated the example in my post.

**Comment by rre36** [comment]
Extended my post even more with a potential way to incorporate fixed size render targets/images with this method.

**Comment by bobcao3** [comment]
Ping-pong buffers will not be needed if this is going to happen, we should be declaring them as separate images if needed
